solver/llbsolver/solver.go

```diff
@@ -587,12 +591,20 @@ func runCacheExporters(ctx context.Context, exporters []RemoteCacheExporter, j *
 						})
 						return err
 					}); err != nil {
-						return prepareDone(err)
+						if !exp.IgnoreError {
```

Comment:
In here you can still mark the step as errored and it will so as such in progressbar. Or write logs to it etc. As long as you return nil in the end, that would not mark the build itself as failing.

Reply:
i've tried to just use `prepareDone(err)` without any return, but it seems without return, the export will still be reported as successfull, while it's not true. And ofc if I use `prepareDone(err)`, the build itself will fail.

Is it OK this way then ?

Reply:
Not sure I understand. With the latest code iiuc the behavior if `ignore-error` is set is that for user everything finishes cleanly(return nil) and the only difference is that a daemon log is printed.

with `prepareDone(err); return nil` the behavior should be that user sees the error in progressbar but the error does not fail the build.

Also, I'm not sure why the `prepareDone(nil)` is before the `Finalize()` call. I think this might be a bug.

Reply:
Yes I believe too that `prepareDone(nil)` before the `Finalize()` is strange.

Yes I thought too that `prepareDone(err)` should display the error on the progress bar, but it wasn't the case.
I've also seen a `// TODO: set error on status` in OneOff function from `util/progress/progress.go`, I've tried to set the Error on Status then, but still didn't find any error on the progress bar during my tests.
Finally, I've edited the function  OneOff to Display the Error as an ID if it exists, and then I can see it on the progress bar

You can see below the test:

```
#9 exporting cache to s3
#9 preparing build cache for export
#9 preparing build cache for export 3.6s done
#9 failed to check file presence in cache: operation error S3: HeadObject, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Head "http://fake-url:9000/my-bucket/blobs/sha256%3A0f9e0b6bb8dc7931f5eaa00d90788f9c80386930b56472726aa1874caa212535": dial tcp: lookup fake-url on 192.168.179.1:53: no such host 3.6s done
#9 DONE 3.6s
+ echo S3 Checks ok
S3 Checks ok
+ docker rmi moby/buildkit:s3test
Untagged: moby/buildkit:s3test
Deleted: sha256:78b17f155de4e360a387e4568a4ff31f928a42b0c37c1f97a3360150b5d92681
Deleted: sha256:108368c5772afe3a6a05d1df867091bb0f5b8f63af25582ee97598214a6879dc
Deleted: sha256:5dcadbc589824476c68e77a9ca041704564de133cfc1af9d93e59f84f87b540f
Deleted: sha256:995803ef80155acdeb9be269ee91c1d5ed7ccca8a010f127f92097f3d95f82c9
Deleted: sha256:634245ec5da3481204bcd22a4571e7c3d1dafcccdedc0370e3f93ae577d2627a
```

Reply:
I was mixing up two functions before. This is called in `inBuilderContext` https://github.com/moby/buildkit/blob/master/solver/llbsolver/solver.go#L573 . The behavior we want is that in https://github.com/moby/buildkit/blob/de7dd7a63328796f7aa43a9d2b06edb94839b012/solver/llbsolver/solver.go#L901 `notifyCompleted` is called with a non-nil error but `inBuilderContext` does not return an error.

Reply:
Oh I see, thanks for the comment
After the modification, we have the right behavior

tested with 
```
buildctl build \
  --progress plain \
  --frontend dockerfile.v0 \
  --local context=/test/test1 \
  --local dockerfile=/test/test1 \
  --export-cache "$default_options,endpoint_url=http://fake-url:9000,ignore-error=true" \
  --output type=local,dest=/tmp/destdir0
```
doesn't return an error because of `ignore-error=true`
```
#9 exporting cache to s3
#9 preparing build cache for export
#9 preparing build cache for export 3.3s done
#9 ERROR: failed to check file presence in cache: operation error S3: HeadObject, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Head "http://fake-url:9000/my-bucket/blobs/sha256%3A6cd68ba108011b0eb1ab7238ed7522a1b2c04a1ef2c00d63426f80ae16438847": dial tcp: lookup fake-url on 192.168.179.1:53: no such host
------
 > exporting cache to s3:
------
S3 Checks ok
+ echo S3 Checks ok
+ docker rmi moby/buildkit:s3test
Untagged: moby/buildkit:s3test
Deleted: sha256:8c64e76d7d0630b0bf9d128f075ba404b7037ec1c52835aab3d0ee09c3dc9eb3
Deleted: sha256:0e26b15e6eedbdf04961feeb5ead3d44b8dd7f7dab5cc6d7a0e5d72643bc2111
```

tested with 
```
buildctl build \
  --progress plain \
  --frontend dockerfile.v0 \
  --local context=/test/test1 \
  --local dockerfile=/test/test1 \
  --export-cache "$default_options,endpoint_url=http://fake-url:9000" \
  --output type=local,dest=/tmp/destdir0
```
return an error as expected because we don't use `ignore-error`
```
#9 exporting cache to s3
#9 preparing build cache for export
time="2022-12-22T09:39:39Z" level=error msg="/moby.buildkit.v1.Control/Solve returned error: rpc error: code = Unknown desc = failed to check file presence in cache: operation error S3: HeadObject, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Head \"http://fake-url:9000/my-bucket/blobs/sha256%3Ab3d9ec013fed62f570d83dfb79b64d6a98d66dae019f281eac5dddf11a4ea991\": dial tcp: lookup fake-url on 192.168.179.1:53: no such host"
#9 preparing build cache for export 1.0s done
#9 ERROR: failed to check file presence in cache: operation error S3: HeadObject, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Head "http://fake-url:9000/my-bucket/blobs/sha256%3Ab3d9ec013fed62f570d83dfb79b64d6a98d66dae019f281eac5dddf11a4ea991": dial tcp: lookup fake-url on 192.168.179.1:53: no such host
------
 > exporting cache to s3:
------
error: failed to solve: failed to check file presence in cache: operation error S3: HeadObject, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Head "http://fake-url:9000/my-bucket/blobs/sha256%3Ab3d9ec013fed62f570d83dfb79b64d6a98d66dae019f281eac5dddf11a4ea991": dial tcp: lookup fake-url on 192.168.179.1:53: no such host
```